**It is assumed that you have installed Python and it is up and running.** 

### PART 1

Python is -

 1. high-level language. JS is also high level language. C, C++ are low level language. 
 
 
 2. a language with dynamic typing. C, C++ and Java are statically typed languages. JS is also dynamically typed language.
 
 
 3. OOP language
 
 
 4. a langauge where (almost) everything is an object. 
 
 
 5. a garbage collected language. C is not a garbage collected language.  
 
 
I've mentioned what kind of langauge Python is but I am mostly skipping the explanation of what these things means. They will be made clear in future. 

**Python as a calculator**

In [9]:
2+2*4, 23/11, 23%11, 23//11, 2**3 # multiplication is higer up than addition

(10, 2.090909090909091, 1, 2, 8)

In [56]:
1/0 #error 

ZeroDivisionError: division by zero

In [46]:
2-2, 3-4

(0, -1)

In [47]:
2 == 2

True

`//` is called floored division. Also referred to as integer division. The resultant value is a whole integer, though the result’s type is not necessarily int. The result is always rounded towards minus infinity: `1//2` is `0`, `(-1)//2` is `-1`, `1//(-2)` is `-1`, and `(-1)//(-2)` is `0`.

`%` is called modulo operator. 

The `%` (modulo) operator yields the remainder from the division of the first argument by the second. The numeric arguments are first converted to a common type. A zero right argument raises the `ZeroDivisionError` exception. The arguments may be floating point numbers, e.g., `3.14%0.7` equals `0.34` (since `3.14` equals `4*0.7 + 0.34`.) **The modulo operator always yields a result with the same sign as its second operand (or zero); the absolute value of the result is strictly smaller than the absolute value of the second operand.** [Source](https://docs.python.org/3/reference/expressions.html#binary-arithmetic-operations)

Python defines `pow(0, 0)` and `0 ** 0` to be `1`, as is common for programming languages.

In [4]:
22%(-3), 22//-3

(-2, -8)

In [341]:
print(22//10)
print(-22//10)
print(-22%10)

2
-3
8


#### Comparison operators

In [6]:
2 < 3, 2 <=1

(True, False)

In [8]:
3 >= 2, 4 > 5

(True, False)

In [7]:
2 != 1

True

In [49]:
2 == 3, 2 ==2 

(False, True)

#### Literals (or fixed values or raw data)

 1. String literals
 2. Numeric literals
 3. Boolean literals

In [21]:
'mayank', "mayank" #constant values, string

('mayank', 'mayank')

In [12]:
2 #constant values, integer

2

In [16]:
2.33 #constant values, floating point

2.33

In [28]:
4+5j  #complex number


(4+5j)

In [29]:
True  #boolean literal

True

#### Python Keywords

Below is the list of Python keywords. They have special meaning and purposes. Don't use them as variable name. 

In [20]:
import keyword  
print(keyword.kwlist) #Python uses these words. 

['False', 'None', 'True', 'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


#### Variables 

In the context of Python, a variable is just a pointer which points to an object. Carefully note that, **a variable is not the object but a pointer to an object it is currently pointing to.** 

A variable, per se, does not have any type (checked by `type` function) or identity (checked by `id` function which returns the memory location) or any value. When you use `type` function or `id` function on a variable, you get the type and id of the object pointed by that variable.

Also note that you can always make a variable to point to any other object. In that case, the object earlier pointed by this variable is deleted (unless pointed by any other variable).





In [25]:
#variable, it can be assigned some values, values can be changed

a = 2
a, type(a), id(a)

(2, int, 739586042192)

In [26]:
a = 3.33  #now 'a' points to different value
a, type(a), id(a)

(3.33, float, 739675497456)

When we executed `a = 2`, Python created an object with value `2` of type `int` at the memory address `739586042192`. And to make this accessible to us, gave it a name `a`.

Then we ran `a = 3.33`, Python created another object with value `3.33` of type `float` at the memory address `739675497456` with name `a`.

You can see that in the former case, `a` was pointing to the object residing at memory address `739586042192` but then we made it to point to another object of different type residing at memory address `739675497456`.

This also explains what we mean when we say  "Python is dynamically typed language".

Now the question arises - what happens to the object earlier pointed by `a`? Since that object is no longer accessible (that is not pointed by any variable), Python deletes it and free up the memory claimed by it. (Python is garbage collected language).

So you can see that variables themselves do not have any type, identity or value. Only objects pointed by them do. 

Also note that multiple variables can point to a single object. See below - 





In [27]:
a = b = 3 

id(a), id(b)

(739586042224, 739586042224)

In [29]:
a = 4

print("a is now :- ",a)
print("b is now :- ", b)

a is now :-  4
b is now :-  3


In [30]:
id(a), id(b)

(739586042256, 739586042224)

With `a = b = 3`, we created a **single** `int` object with value `3` at memory address `739586042224` and this object is pointed to by 2 variables `a` and `b`.

Then with `a = 4`, we created another `int` object with value `4` at memory location `739586042256`. After this, earlier object can no longer be accessible through variable `a`. Since earlier object is still pointed to by `b`, it isn't deleted by Python's garbage collection mechanism. 

See another example.



In [31]:
a = 'Peter'
b = a

id(a), id(b)


(739663577136, 739663577136)

In [32]:
b = 4
print(a), id(a), id(b)

Peter


(None, 739663577136, 739586042256)

Note that before we can use any variable, we have to assign some value to them. It shouldn't come as a surprise. After all, variable has to point to some object. Without object, variable has no meaning. 

In [54]:
n  #error

NameError: name 'n' is not defined

**This section is almost completely copied from Whirlwind Tour of Python because it was so good.**


Python Variables Are Pointers - Assigning variables in Python is as easy as putting a variable name to the left of the equals (``=``) sign:
```python
# assign 4 to the variable x
x = 4
```

This may seem straightforward, but if you have the wrong mental model of what this operation does, the way Python works may seem confusing. We'll briefly dig into that here.

In many programming languages, variables are best thought of as containers or buckets into which you put data.
So in C, for example, when you write
```
// C code
int x = 4;
```
you are essentially defining a "memory bucket" named ``x``, and putting the value ``4`` into it. In Python, by contrast, variables are best thought of not as containers but as pointers. So in Python, when you write
```python
x = 4
```
you are essentially defining a *pointer* named ``x`` that points to some other bucket containing the value ``4``.
Note one consequence of this: because Python variables just point to various objects, there is no need to "declare" the variable, or even require the variable to always point to information of the same type! This is the sense in which  people say Python is *dynamically-typed*: variable names can point to objects of any type. So in Python, you can do things like this:
```python
x = 1         # x is an integer
x = 'hello'   # now x is a string
x = [1, 2, 3] # now x is a list
```
While users of statically-typed languages might miss the type-safety that comes with declarations like those found in C,
```
int x = 4;
```
this dynamic typing is one of the pieces that makes Python so quick to write and easy to read. There is a consequence of this "variable as pointer" approach that you need to be aware of.

If we have two variable names pointing to the same *mutable* object, then changing one will change the other as well! For example, let's create and modify a list:

```python
x = [1, 2, 3]
y = x
```
We've created two variables ``x`` and ``y`` which both point to the same object. Because of this, if we modify the list via one of its names, we'll see that the "other" list will be modified as well:

```python
print(y)       #output – [1,2,3]
x.append(4)    # append 4 to the list pointed to by x
print(y)       #output – [1,2,3,4] y's list is modified as well!
```

This behavior might seem confusing if you're wrongly thinking of variables as buckets that contain data. But if you're correctly thinking of variables as pointers to objects, then this behavior makes sense.
Note also that if we use "``=``" to assign another value to ``x``, this will not affect the value of ``y`` – assignment is simply a change of what object the variable points to:

```python
x = 'something else'
print(y)  # y is unchanged
```
Again, this makes perfect sense if you think of ``x`` and ``y`` as pointers, and the "``=``" operator as an operation that changes what the name points to.

Numbers, strings, and other *simple types* are immutable: you can't change their value – you can only change what values the variables point to. So, for example, it's perfectly safe to do operations like the following:
```python
x = 10
y = x
x += 5  # add 5 to x's value, and assign it to x
print("x =", x) #x=15
print("y =", y) #y=10
```
When we call ``x += 5``, we are not modifying the value of the ``10`` object pointed to by ``x``; we are rather changing the variable ``x`` so that it points to a new integer object with value ``15``. For this reason, the value of ``y`` is not affected by the operation.


**END**

In [34]:
a = None
b = 1
type(b), a.__class__

(int, NoneType)

#### Python Data types

 - `None`: This type has single value `None`. 
 - Numerics types: They are immutable.
   - Integers: 2 types of integers: Integers and Booleans 
     - `int` type for any integer. They represent all integers
     - `bool` type. Only two possible values `True` representing `1` and `False` representing `0`.
   - floats: `float` type. Represents floating value numbers such as 2.3
   - complex numbers: `complex` types. Complex numbers such as 2+3j
 - String: `str` type. Sequence of characters. These are immutable
 - Lists: `list` Lists contain orderly collection of other arbitrary Python objects. They are mutable.
 - Tuples: `tuple` Just like lists but immutable
 - Dictionaries: `dict` type. Mapping objects. Collection of `key:value` pairs. Mutable. 
 - Sets: `set` type. These represent unordered, finite sets of unique, immutable objects. They are immutable. 
 

These are the most used data types. Other data types are also available but they are rarely used hence not mentioned here.   

See below for examples of each types.


In [33]:
none_value = None
num = 1        #integer, number
flt = 2.2      #float, also a number
comp = 2+7j    #complex number
boo = True     #boolean value
st = 'abc'     #string
st1 = "afsdf"  #also a string
st2 = "111"    #also a string

foo = [1,2,'a', 'mayank', True] #a list
bar = (1,2,'a', 'mayank', True) #a tuple
baz = {'a': 1, 'b': False, 'c': 'xyz'}  #a dictionary
bam = {1,'a','b'} #a set

Good news is that these are the only data types we'll work with. 



#### Built-in types

Above we mentioned some built-in data types. Let's check `type` of each variable we defined above. 

In [40]:
type(num), type(flt), type(comp), type(boo)

(int, float, complex, bool)

In [41]:
type(st), type(st1), type(st2)

(str, str, str)

In [42]:
type(foo), type(bar), type(baz), type(bam)

(list, tuple, dict, set)

There are many other types available in Python, however, they are not necessarily data types. For example, take a built-in function `sum()` and check its type. 

In [43]:
type(sum)

builtin_function_or_method

Or a class:

In [44]:
class Test():
    pass

type(Test)

type

In [45]:
test = Test()

type(test)

__main__.Test

Above two example may stump you but don't worry. Their types aren't of immediate concern. Important point here is that there are many other possible types and they might not necessarily be data types. 

### PART 2

In this part, we'll look into numerics types, strings, lists, tuples, dictionaries and sets. 

types|iterables|sequence|mutability
------|-----|------|-------
numeric types|No|No|Immutable
string|Yes|Yes|Immutable
List|Yes|Yes|Mutable
Tuple|Yes|Yes|Mutable
Dictionaries|Yes|No|Mutable
Sets|Yes|No|Mutable

In very simple terms -

- If an object is iterable, you can iterate over its elements using `For` loop.
- If an object is sequence, you can access its elements by indexing (using `object[i]` notation)
- If an object is mutable, you can change its content without changing its identity.  

#### Numbers

In [312]:
a =  4

[i for i in dir(a) if not i.startswith('_')]

['as_integer_ratio', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']

In [12]:
some_float = 3433.121345

some_float.as_integer_ratio()

(7549513676786961, 2199023255552)

In [14]:
some_num = 121
some_num.bit_length(), bin(121)

(7, '0b1111001')

In [16]:
comp = 3+4j
comp.conjugate(), comp.imag, comp.real

((3-4j), 4.0, 3.0)

In [18]:
xy = 5
xy.denominator, xy.numerator

(1, 5)

**Note about boolean values**

In numeric contexts (for example when used as the argument to an arithmetic operator), they behave like the integers 0 and 1, respectively.

In [334]:
True == 1, True + 1

(True, 2)

In [333]:
2 - False

2

In [342]:
issubclass(bool,int)

True

In [343]:
isinstance(True, int)

True

In [6]:
bool.__bases__, int.__subclasses__()
#bool class inherits from `int` class

((int,),
 [bool,
  <enum 'IntEnum'>,
  <enum 'IntFlag'>,
  sre_constants._NamedIntConstant,
  subprocess.Handle])

Complex numbers are written in the form `a+bj`. If  you want to use simply `j`, write it as `1j`.  

In [336]:
1j * 1j

(-1+0j)

Python can also deal with `NaN` ('Not a Number'), $-\infty$ and $\infty$. You simply pass the strings 'NaN', 'nan' for `NaN` and 'inf' or '-inf' for $\infty$ or $-\infty$.

In [24]:
a = float('NaN')
a

nan

In [35]:
float('inf')

inf

In [22]:
3 > float('-inf')

True

In [26]:
float('-inf') == float('-inf')


True

In [53]:
2**1000 > float('inf')

False

Note that `NaN` doesn't equal to anything, not even to itself so `==` can't be used to test equality.

In [27]:
float('NaN') == float('NaN')

False

In [54]:
a = float('NaN')

float

#### Strings

In [85]:
a = '' #empty string
b = str() #also an empty string

type(a),type(b)

(str, str)

Strings can be concatenated -

In [250]:
'Sam' + 'wise'

'Samwise'

In [251]:
name = 'Sam'
print(name + 'wise') #name is a variable which holds a string

Samwise


Below is the list of string methods. We'll talk about methods later but for the time being, just keep in mind that every object has methods which can be used by `.` notation.  

In [88]:
%pprint      
[i for i in dir('') if not '_' in i] #this is list comprehension but don't bother about it just yet

Pretty printing has been turned OFF


['capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']

In [90]:
a = 'mayank'
a.capitalize()  #this is how you use method via dot notation. 

'Mayank'

You are supposed to explore every string methods. 

We can access each character of a string by subscription method. Remember that in Python, indexing starts from 0. That means first element will actually be 0th element in programming terminology. 

In [96]:
name = 'mayank'
name[0]         #subscription method

'm'

In [97]:
len(name)

6

In [151]:
name[::-1]  #string reversal through slicing 

'knayam'

In [182]:
name[:2]

'ma'

In [152]:
for i in name:
    print(i)

m
a
y
a
n
k


In [153]:
list(name)

['m', 'a', 'y', 'a', 'n', 'k']

In [155]:
sorted('peter')

['e', 'e', 'p', 'r', 't']

In [161]:
'i' in name, 'm' in name

(False, True)

white space characters such as `\n` are treated as a character 

In [238]:
print("maya\nk")

maya
k


To 'escape' this, you use `\` before anything you want to escape. For example, if you want to print `\n` as such, you escape it as shown below -

In [239]:
print("maya\\nk")

maya\nk


Alternatively, you can use *raw strings* by prefixing `r` before first quote mark - 

In [245]:
print(r'maya\nk')

maya\nk


See also - 

In [240]:
print("He said, 'Hi'")

He said, 'Hi'


In [241]:
print("He said, "Hi""") #error

SyntaxError: invalid syntax (<ipython-input-241-cf09fba1d82f>, line 1)

In [243]:
#use escaping 
print("He said, \"Hi\"")

He said, "Hi"


Multi-line string

In [248]:
print('''
First enter your name
Then, your age
''')


First enter your name
Then, your age



In [249]:
print('''\
First enter your name
Then, your age
''')

First enter your name
Then, your age



Carefully note that in previous example there was a blank line while in second example, newline character was omitted because we escaped this. 

Strings are **immutable** which means, once created, you can not change the value of string. More on that later. 

In [207]:
xyz = 'mayank'
xyz[1] = 'z'    #trying to change first 'a' to 'z', error

TypeError: 'str' object does not support item assignment

#### lists


In [159]:
list1 = list()  #empty list
list2 = []      #empty list
list3 = [1,2]   #non-empty list

list1,list2,list3         

([], [], [1, 2])

In [164]:
[i for i in dir(list1) if not '_' in i]

['append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']

These are the methods for list. Most of them are pretty intuitive. However, you are supposed to practice them through examples. Apart from these methods, some built-in functions such as `len()`, `sorted()` can also be applied on a list. 

In [165]:
names = ['Mayank', 'Peter', 'Rajat']

len(names)

3

In [166]:
sorted(names)

['Mayank', 'Peter', 'Rajat']

In [167]:
all(names), any(names)

(True, True)

In [169]:
for i in reversed(names):
    print(i)

Rajat
Peter
Mayank


In [172]:
min(names), max(names)

('Mayank', 'Rajat')

Since elements of a list are ordered, we can access them through indexing

In [174]:
names[0]

'Mayank'

In [175]:
names[-1] # -1 means indexing starts from last element

'Rajat'

Lists can be sliced through `list[:start:stop:step]` method. `start` defaults to 0, `stop` defaults to length of list minus 1 and `step` defaults to 1.

In [178]:
values  = [3,4,3,7,8,9,5]
values[::], values[:2]

([3, 4, 3, 7, 8, 9, 5], [3, 4])

However, if `step` is negative number, `start` defaults to length of list minus 1, `stop` defaults to 0. 

In [180]:
values[::-1] #list reversal

[5, 9, 8]

In [206]:
values[:3:-1], values[6:3:-1] #slicing from end. Careful here

([5, 9, 8], [5, 9, 8])

In [200]:
values[5:3:-1]

[9, 8]

In [186]:
11 in values, 4 in values 

(False, True)

In [190]:
sum(values)

39

List can be replicated by `*` operator

In [191]:
names*3

['Mayank', 'Peter', 'Rajat', 'Mayank', 'Peter', 'Rajat', 'Mayank', 'Peter', 'Rajat']

Carefully observe the following examples

In [192]:
names

['Mayank', 'Peter', 'Rajat']

In [194]:
names1 = names
names is names1  #both are same objects

True

In [195]:
names2 = list(names)

names2 is names  #both are different objects

False

Note, however, that even though `names` and `names2` refer to 2 different lists, their contents still refer to same objects. This is **shallow copying**. Only outermost content is copied. 

In [319]:
names[1] is names2[1]

True

In [197]:
names3= names[::]

names3 is names  #both are different lists

False

In [320]:
names[1] is names3[1]

True

Lists are **mutable** which means you can change their content. 

In [208]:
random_list = [2,3,4,5]
print(id(random_list))
random_list[3] = 11
print(random_list)
print(id(random_list))

404477705856
[2, 3, 4, 11]
404477705856


You can see that the content of list has changed but its identity (value of `id()` function) hasn't.

**List comprehensions** provide a concise way to create lists. This basic syntax, then, is `[expr for var in iterable optionally 0 or more if or for clause]`, where `expr` is any valid expression, `var` is a variable name, and `iterable` is any iterable Python object.

In [260]:
[i**2 for i in range(5)]

[0, 1, 4, 9, 16]

In [261]:
#this listcomp combines the elements of two lists if they are not equal:

[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]

[(1, 3), (1, 4), (2, 3), (2, 1), (2, 4), (3, 1), (3, 4)]

In [262]:
foo = [[1,2],[3,4]]
[f for bar in foo for f in bar]


[1, 2, 3, 4]

In [263]:
[(i, j) for i in range(2) for j in range(3)]


[(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]

Notice that the second `for` expression acts as the interior index, varying the fastest in the resulting list. This type of construction can be extended to three, four, or more iterators within the comprehension, though at some point code readibility will suffer!

In [273]:
colors = ['red','blue','green' ]
toys = ['car','ball','duck']

[[color, toy] for color in colors for toy in toys]

[['red', 'car'], ['red', 'ball'], ['red', 'duck'], ['blue', 'car'], ['blue', 'ball'], ['blue', 'duck'], ['green', 'car'], ['green', 'ball'], ['green', 'duck']]

#### `sorted()` vs `list.sort()`

In [264]:
some_list = [2,1,5,3]
sorted(some_list)

[1, 2, 3, 5]

In [265]:
some_list

[2, 1, 5, 3]

In [267]:
some_list.sort()
some_list

[1, 2, 3, 5]

As you can see, `sorted()` function doesn't change the original list but `list.sort()` does.

#### `list.append()` vs `list.extend()`

In [268]:
some_list

[1, 2, 3, 5]

In [269]:
some_list.append([1,2]) #this method takes an object as argument append the entire object

In [270]:
some_list

[1, 2, 3, 5, [1, 2]]

In [271]:
some_list.extend([11,12]) #this method takes an iterable as an argument and append each element of that iterable

In [272]:
some_list

[1, 2, 3, 5, [1, 2], 11, 12]

**Aside**

We have learned to use list methods like this - `some_list.sort()`. However, there is another way. You can also do `list.sort(some_list)`. Although, we haven't introduced OOP yet, but keep in mind that `some_list` is an instance of a `list` class and former is known as bound method call while later is unbound method call. [See this SO post](https://stackoverflow.com/questions/6083090/in-python-whats-difference-between-some-string-lower-and-str-lowersome-stri).

The same is true for string methods which we'll encounter in later section. `some_string.lower()` produces the same result as `str.lower(some_string)`.

#### A List Gotcha



In [1]:
a = [[None]*2]*3
a


[[None, None], [None, None], [None, None]]

In [2]:
id(a), id(a[0])

(739675491904, 739675491200)

In [3]:
id(a[0]) == id(a[1]) == id(a[2])

True


The lists `a[0]`,`a[1]` and `a[2]` are pointing to same list. They all have same `id()` value.

In [4]:
id(a[0][0]) == id(a[1][0]) == id(a[2][0])

True

In [5]:
a[0][0] =1
a

[[1, None], [1, None], [1, None]]

In [6]:
id(a), id(a[0])

(739675491904, 739675491200)

We tried to change only one element and here 3 elements are changed!

The lists `a[0]`, `a[1]`, `a[2]` are STILL pointing to same list. They all STILL have same `id()` value.
The reason behind above is that replicating a list with `*` doesn’t create copies, it only creates references to the existing objects. The `*3` creates a list containing 3 references to the same list of length two. Changes to one row will show in all rows, which is almost certainly not what you want. Try below instead - 


In [7]:
a = [None] * 3 # a = [None, None, None]
for i in range(3):
    a[i] = [None] *2 #each time a new list will be inserted

a    

[[None, None], [None, None], [None, None]]

In [8]:
a[0][0] = 1
a


[[1, None], [None, None], [None, None]]

#### One of many ways for checking for empty list

In [9]:
l = []
l and 1


[]

In [10]:
l = []
if not l:      # not l evaluates to true as empty list evaluates to false
    print("list is empty")
else:
    print(l)

list is empty


#### Inserting an element in a sorted list and ensure that list remains sorted

In [20]:
a = [1,2,4,5]

import bisect
bisect.insort(a, 3)

In [21]:
a

[1, 2, 3, 4, 5]

#### Tuples

Tuples are similar to lists except for only one difference. Unlike lists, tuples are immutable.

In [52]:
a = ()      #empty tuple
b = (1,)    #note the use of comma. Without comma 'b' will be a number
c = (1)     #not a tuple, but an integer
d = tuple() #tuple function used here
e = 1,2,'a' #also a tuple

type(a), type(b), type(c), type(d), type(e)

(tuple, tuple, int, tuple, tuple)

Tuples have following associated methods - 


In [214]:
[i for i in dir(a) if not '_' in i] #'a' is a tuple from above example

['count', 'index']

You can see tuples don't have any method such as `.append()` which can change its content. This shows the immutability of tuples. 

We saw that `a = list(b)` and `a = b[::]` create 2 different lists. Let's see see what happens in case of tuples. 

In [321]:
tuple1 = (1,2,[1])
tuple2 = tuple(tuple1)

tuple1 is tuple2

True

In [323]:
tuple3 = tuple1[::]
tuple3 is tuple1

True

We see that in case of tuples, we get same tuple. In general,

- **`b = a`** - both refers to same object, irrespective of their types.
- **`b = constructor(a)`** - if type mutable, both refer to different object otherwise same object
- **`b = a[:]`** - if type mutable, both refer to different object otherwise same object.

#### Dictonaries

Dictonaries are objects which have collections of `key:value` pairs. 

In [220]:
dict1 = {}                                     #empty dictionary
dict2 = dict()                                 #empty dictionary created by using dict() function 
dict2 = {'name': 'Peter', 'age': 35}           # 'name' & 'age' are keys and 'Peter' & 35 are their respective values


In [259]:
dict([('sape', 4139), ('guido', 4127), ('jack', 4098)])

{'sape': 4139, 'guido': 4127, 'jack': 4098}

If keys are all strings,  it is sometimes easier to specify pairs using keyword arguments: - 


In [254]:
dict(name = 'Peter', Country ='USA')m

{'name': 'Peter', 'Country': 'USA'}

Keys can be of any immutable type. If tuples are used as keys, make sure tuples don't contain any mutable object such as a list. 

In [255]:
{(1,2):'foo'}

{(1, 2): 'foo'}

In [256]:
{(1,[2]): 'bar'}  #error

TypeError: unhashable type: 'list'

**Dict comprehension** can also be used to create a dictionary - 

In [258]:
{x:x**2 for x in [2,3,4]}

{2: 4, 3: 9, 4: 16}

Below is the list of all the methods which can be applied on a dictionary. You are supposed to practice all of them. 

In [313]:
[i for i in dir(dict1) if not i.startswith('_')] 

['clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']

In [14]:
c = {}
help(c.setdefault)

Help on built-in function setdefault:

setdefault(key, default=None, /) method of builtins.dict instance
    Insert key with a value of default if key is not in the dictionary.
    
    Return the value for key if key is in the dictionary, else default.



In [233]:
len(dict2)

2

In [234]:
'name' in dict2  #you can check if some key is present in dictionary

True

In [235]:
'Peter' in dict2 #there is no 'Peter' key. values can't checked for presence in dictionary 

False

In [236]:
#you can loop over dictionary keys

for i in dict2:
    print(i)

name
age


#### Sets


In [229]:
set1 = set()      #empty set using set() function
set2 = {1,2,3,4}  #set
no_set1 = {}      #not a set
list5 = [1,2,3,1] 
set3 = set(list5) ##will prepare a set by removing duplicate items

print(set3)

{1, 2, 3}


In [315]:
[i for i in dir(set1) if not i.startswith('_')] 

['add', 'clear', 'copy', 'difference', 'difference_update', 'discard', 'intersection', 'intersection_update', 'isdisjoint', 'issubset', 'issuperset', 'pop', 'remove', 'symmetric_difference', 'symmetric_difference_update', 'union', 'update']

In [230]:
4 in set3, 3 in set3

(False, True)

In [231]:
len(set3)

3

In [232]:
set3[1]  #indexing doesn't work with sets

TypeError: 'set' object is not subscriptable

### Some code snippets


In [326]:
#a list of lists is given. You want to create a list consisting of unique items.

lol = [['a','b','c'],['b','c','d'],['d','e']]
a = set().union(*lol)
list(a)

['b', 'e', 'c', 'd', 'a']

In [327]:
#merging two dictionaies into one

a = {'b':3, 'c':4}
b = {'b':7,'d':9}
merged = {**a, **b} #for 3.5++
merged              #Notice for key b, lattermost value was used.

{'b': 7, 'c': 4, 'd': 9}

In [328]:
# Combining 2 lists - One list as keys and items from other list are grouped as values

a = ['jan','jan', 'feb', 'mar', 'feb']
b = [1,2,3,4,5]
c = {}
for x, y in zip(a,b):
    c.setdefault(x,[]).append(y)
    
c    

{'jan': [1, 2], 'feb': [3, 5], 'mar': [4]}

### PART 3

Before we move to loops and control flows, we need to understand **Truth value testing.**

Any object can be tested for truth value, for use in an `if` or `while` condition or as operand of the Boolean
operations. Every object is either `True` or `False`. The general intuition is that `False` denotes zero or emptiness while `True` denotes some value. 

Here are most of the built-in
objects considered false:

 - constants defined to be false: `None` and `False`.
 - zero of any numeric type: `0`, `0.0`, `0j`, `Decimal(0)`, `Fraction(0, 1)`
 - empty sequences and collections: `''`, `()`, `[]`, `{}`, `set()`, `range(0)`

Operations and built-in functions that have a Boolean result always return `0` or `False` for false and `1` or
`True` for true, unless otherwise stated. 

(Important exception: the Boolean operations `or` and `and` always return one of their operands. Explained later)

We can use built-in `bool()` function to test the truth value. See following examples - 

In [68]:
a = [1] #non-empty list and truth value is True
b = []  #empty list, hence truth value is False
c = {}  #empty dictionary, False
d = ''  #empty string
e = 'mayank' #non-empty string, True
f = True #always 
g = False #always False
h = None #False
i = 0 #zero is False

bool(a),bool(b),bool(c),bool(d),bool(e),bool(f),bool(g),bool(h),bool(i)

(True, False, False, False, True, True, False, False, False)

#### Boolean Operations - **`and`, `or`, `not`**

These are the Boolean operations, ordered by ascending priority:

|Operation            |Result|Notes|
|---------------------|------|-----|
|`x or y`|if `x` is false, then `y`, else `x`|1|
|`x and y`|if `x` is false, then `x`, else `y`|2|
|`not x`|if `x` is false, then `True`, else `False`|3|


1. This is a short-circuit operator, so it only evaluates the second argument if the first one is false.
2. This is a short-circuit operator, so it only evaluates the second argument if the first one is true.
3. `not` has a lower priority than non-Boolean operators, so `not a == b` is interpreted as `not (a == b)`, and `a == not b` is a syntax error.

In [69]:
print(3 and 5)     #'x and y' operation. 'x' is not false so it returns 'y', that is, 5. 
print(5 and 3)
print(5 and False)
print(False and 5)
print(None and 5)
print(0 and 5)

5
3
False
False
None
0


In [70]:
print(3 or 5)      # 'x or y' operation. if 'x' false, 'y' is output. Otherwise 'x' is output. 
print([] or 3)
print(0 or 3)
print(3 or 0)
print(False or 3)
print(0 or False)

3
3
3
3
3
False


In [71]:
1 and 0 or 3 #left to right evaluation

3

While we are at it, let us also look into `&` and `|` operators. They are similar to `and` and `or` but they operate on bits of objects rather than on objects themselves. 

In [77]:
3 and 122, 3 & 122  #see the difference


(122, 2)

In [83]:
print(bin(3))
print(bin(122))
print(bin(3 & 122))


0b11
0b1111010
0b10


In [287]:
import numpy as np 
A = np.array([1, 0, 1, 0, 1, 0], dtype=bool)
B = np.array([1, 1, 1, 0, 1, 1], dtype=bool)
A | B

array([ True,  True,  True, False,  True,  True])

In [288]:
A or B #error

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

#### Python ternery operator

In [289]:
a = 1
b = 2
1 if a > b else -1


-1

In [70]:
a,b,c = [1,2,3]
a < b == c


False

In [72]:
a < b and b < c

True

In [292]:
a,b,c = [2,3,3]
a < b or b < c

True

In [293]:
d = a or b or c
d

2

In [294]:
d =  a and b and c
d


3

### `if-elif-else` conditions

With `if`, `elif` and `else` are optional. 

Syntax is:

```Python
if <exp>:         
    <code block>
elif <exp>:
    <code block>
else:
    <code block>
```

    

In [68]:
if 1:
    print('Hi')
    
    

Hi


In [69]:
if []:
    print('Nothing is printed')

In [41]:
if False:
    print("nothing will be printed")

In [42]:
if True:
    print("HELLO!")

HELLO!


In [284]:
value  =  5

if value < 6:     # 5 < 6 evaluates to True
    print('Fine')

Fine


In [285]:
age = int(input("Enter age:\n"))

if age < 40:                  #55 < 40 evaluates to False so 'if' clause won't be executed
    print("You are young")
else:             
    print("No more young!")

Enter age:
55
No more young!


In [286]:
value =  3

if value in [2,3,4,5,6]:
    print("value is in container")

value is in container


In [40]:
nums = [2,3,4]

for i in nums:
    if i%2==0:
        print('{} is even number'.format(i))
    else:
        print('{} is odd number'.format(i))
    

2 is even number
3 is odd number
4 is even number


In [55]:
marks = -90

if marks >=0 and marks <= 50:
    print("Your marks are not good")

elif marks > 50 and marks <= 75:
    print("Your marks are good")
    
elif marks > 75 and marks <= 100:     #using 'else'clause here can cause problems
    print("Your marks are excellent!")

In codes where you are using `elif` clause(s), be careful that the condition you are testing should be =`True` in exactly one of `if-elif-else` clauses. 

Can you see if there is any problem in following code?

```Python
marks = 90

if marks >=0 and marks <= 50:
    print("Your marks are not good")

elif marks > 50 and marks <= 75:
    print("Your marks are good")
    
else:     
    print("Your marks are excellent!")
```



In [49]:
#LOL!!

marks = -90

if marks >=0 and marks <= 50:
    print("Your marks are not good")

elif marks > 50 and marks <= 75:
    print("Your marks are good")
    
else:     #using 'else'clause here can cause problems
    print("Your marks are excellent!")

Your marks are excellent!


### `for` loop

Syntax is

```Python
for i in iterable:
    #do something with i
```    
Runs a block of code for each element in a given sequence.


In [274]:
values = [1,2,3,4]

for i in values:
    print(i*i)

1
4
9
16


In [275]:
for i in range(1,5):
    if i == 2:
        pass
    else:
        print(i)

1
3
4


In [276]:
for i in range(7):
    if i == 4:
        continue
    else:
        print(i)

0
1
2
3
5
6


In [277]:
for i in range(7):
    if i == 4:
        break
    else:
        print(i)

0
1
2
3


### `for - else` 

In [283]:
listing = [1,2,3,6,8]
for i in listing:
    if i == 6:
        print(i)
        break    
else:
    print('not there')  


6


In above example, `else` is associated with `for` clause and it is executed if `break` is not executed.

### `while` loop

Syntax is 

```Python
while <expression>:
    code block
```

Code block will be executed as long as *expression* is `True`.

In [278]:
a = 0
while a <6: 
    print(a)
    a+=1

0
1
2
3
4
5


In [280]:
#while True introduces infinite loop, can only be terminated via `break` statement

while True:
    name = input("Enter your name:\n")
    if name == "Mayank":
        print("Hi, %s" %name)
        break

Enter your name:
peter
Enter your name:
Mayank
Hi, Mayank


#### `while-else-break` loop

Run example given below. Notice two things. First, notice `else` clause is associated with `while` clause. Second, notice `break` statement and how it affects the execution of `else` clause. `else` clause is run only when `break` statement is not executed. 

In [281]:
a = 1
while (a < 3):
    string =  input("input your name in small letters:\n")
    if string == 'mayank':
        print("Hi %s" %string) 
        break
    a = a + 1

else:
    print("Sorry!")

input your name in small letters:
sam
input your name in small letters:
peter
Sorry!


In [282]:
n = -1
while n < 5:
    n =  n + 1
    if n == 2:
        continue
    print(n)

0
1
3
4
5


`continue` makes program do nothing for current iteration, rather it makes program to jump to next iteration.

### PART 4

This is the part I want to introduce at later part of the course because it introduces design features and some subtle concepts which novices may not be able to fully understand. 

#### Everything is an object 

Every variable you define is an object. Every function, built-in or user-defined, is an object. Every class is an object. Every instance of a given class is an object. This also means you can 'move around' them. You can assign them to variables, you can pass them to functions, functions can return them and so on. 

Also every objects has some characterstics (called attributes) and associated behaviours (called methods). We can access attributes or apply methods on object via `.` notations. 

     obj.attribute
     obj.method()

Every object has a type, an identity and a value. 

**Following section is credited to Jake VanderPlas**

Python is an object-oriented programming language, and in Python everything is an object.
Let's flesh-out what this means. Earlier we saw that variables are simply pointers, and the variable names themselves have no attached type information. This leads some to claim erroneously that Python is a type-free language. But this is not the case!
Consider the following:
```python
x = 4
type(x) #int
x = 'hello'
type(x) #str
x = 3.14159
type(x) #float
```

Python has types; however, the types are linked not to the variable names but *to the objects themselves*. In object-oriented programming languages like Python, an *object* is an entity that contains data along with associated metadata and/or functionality.

In Python everything is an object, which means every entity has some metadata (called *attributes*) and associated functionality (called *methods*). These attributes and methods are accessed via the dot syntax.
For example, before we saw that lists have an ``append`` method, which adds an item to the list, and is accessed via the dot 
("``.``") syntax:
```python
L = [1, 2, 3]
L.append(100)
print(L)     # L = [1, 2, 3, 100]
```
While it might be expected for compound objects like lists to have attributes and methods, what is sometimes unexpected is that in Python even simple types have attached attributes and methods. For example, numerical types have a ``real`` and ``imag`` attribute that returns the real and imaginary part of the value, if viewed as a complex number:  x 
```python
x = 4.5
print(x.real, "+", x.imag, 'i')  # 4.5 + 0.0i
```
Methods are like attributes, except they are functions that you can call using opening and closing parentheses. For example, floating point numbers have a method called ``is_integer`` that checks whether the value is an integer:
```python
x = 4.5
x.is_integer() #False
x = 4.0
x.is_integer() #True
```

When we say that everything in Python is an object, we really mean that *everything* is an object – even the attributes and methods of objects are themselves objects with their own ``type`` information:
```python
type(x.is_integer)       #builtin_function_or_method
```

#### Mutability and Immutability

So far we have seen that strings and tuples are immutable while lists are mutable. Although we might have developed generally accurate understanding of mutable/immutable objects, we better look at it more closely. 

#### Some Excrepts from Python Documentations

##### Data Model

Objects are Python’s abstraction for data. All data in a Python program is represented by objects or by
relations between objects. (In a sense, and in conformance to Von Neumann’s model of a “stored program
computer,” code is also represented by objects.)
Every object has an identity, a type and a value. An object’s identity never changes once it has been created;
you may think of it as the object’s address in memory. The `is` operator compares the identity of two
objects; the `id()` function returns an integer representing its identity.

---
(*following section hints what we mean by mutability*)

An object’s type determines the operations that the object supports (e.g., “does it have a length?”) and also
defines the possible values for objects of that type. The `type()` function returns an object’s type (which is
an object itself). Like its identity, an object’s type is also unchangeable.1
The value of some objects can change. Objects whose value can change are said to be mutable; objects
whose value is unchangeable once they are created are called immutable. (The value of an immutable
container object that contains a reference to a mutable object can change when the latter’s value is changed;
however the container is still considered immutable, because the collection of objects it contains cannot be
changed. So, immutability is not strictly the same as having an unchangeable value, it is more subtle.) An
object’s mutability is determined by its type; for instance, numbers, strings and tuples are immutable, while
dictionaries and lists are mutable.
Objects are never explicitly destroyed; however, when they become unreachable they may be garbagecollected.
An implementation is allowed to postpone garbage collection or omit it altogether — it is a
matter of implementation quality how garbage collection is implemented, as long as no objects are collected
that are still reachable.

---
...Some objects contain references to other objects; these are called containers. Examples of containers are
tuples, lists and dictionaries. The references are part of a container’s value. In most cases, when we talk
about the value of a container, we imply the values, not the identities of the contained objects; however,
when we talk about the mutability of a container, only the identities of the immediately contained objects
are implied. So, if an immutable container (like a tuple) contains a reference to a mutable object, its value
changes if that mutable object is changed.

-----
Types affect almost all aspects of object behavior. Even the importance of object identity is affected in
some sense: for immutable types, operations that compute new values may actually return a reference to
any existing object with the same type and value, while for mutable objects this is not allowed. **E.g., after
`a = 1`; `b = 1`, `a` and `b` may or may not refer to the same object with the value one, depending on the
implementation, but after `c = []; d = []`, `c` and `d` are guaranteed to refer to two different, unique, newly
created empty lists. (Note that `c = d = []` assigns the same object to both `c` and `d`.)**

In [304]:
a = 1
b = 1

a is b

True

In [305]:
a = 300
b = 300

a is b

False

**Note** - For values between [-5,256], `a` and `b` will be identical.

In [301]:
a, b = [],[]
a is b

False

#### Mutability of Tuples

In [337]:
t1 = (1,2,[3,4])

print(t1)
print("id of t1 is %d" %id(t1))
print("id of t1[-1] is %d" %id(t1[-1]))

t1[-1].append(5)

print("now t1 is ", t1)
print("now id of t1 is %d" %id(t1))
print("now id of t1[-1] is %d" %id(t1[-1]))


(1, 2, [3, 4])
id of t1 is 404495510080
id of t1[-1] is 404478705856
now t1 is  (1, 2, [3, 4, 5])
now id of t1 is 404495510080
now id of t1[-1] is 404478705856


Tuples are immutable but above we can see it got 'mutated'. This example shows how tuples can be mutable if one of its elements is of mutable type. In this example, original and final tuple still have same id.

This also illustrates the situation in which the value of a tuple changes as result of changes to a mutable object referenced in it. What can never change in a tuple is the identity of the items it contains. Also – 

In [338]:
t = (1,2,[30,40])
t[2] += [50,60] #id of list won't change, hence tuple will 'mutate'


TypeError: 'tuple' object does not support item assignment

In [339]:
t

(1, 2, [30, 40, 50, 60])

In [340]:
t[2] = t[2] +[90] #won't work as id of list will change which isn't allowed as tuple is immutable

TypeError: 'tuple' object does not support item assignment

#### += Operation on Mutable Types
For any operator "``■``", the expression ``a ■= b`` is equivalent to ``a = a ■ b``, with a slight catch.
For mutable objects like lists, arrays, or DataFrames, these augmented assignment operations are actually subtly different than their more verbose counterparts: they modify the contents of the original object rather than creating a new object to store the result.

In [306]:
s = [1,2,3]
print(id(s))
s = s + [1]
print(id(s))  #id of s changed
s+=[2]
print(id(s))  #id doesn't change

404495335936
404495335680
404495335680


#### Aside: Floating Point Precision

One thing to be aware of with floating point arithmetic is that its precision is limited, which can cause equality tests to be unstable. For example:

In [307]:

float(2**53) == float(2**53) + 1

True

In [308]:
.1 + .2

0.30000000000000004

Why is this the case? It turns out that it is not a behavior unique to Python, but is due to the fixed-precision format of the binary floating-point storage used by most, if not all, scientific computing platforms. All programming languages using floating-point numbers store them in a fixed number of bits, and this leads some numbers to be represented only approximately. We can see this by printing the three values to high precision:


In [309]:
print("0.1 = {0:.17f}".format(0.1))
print("0.2 = {0:.17f}".format(0.2))
print("0.3 = {0:.17f}".format(0.3))

0.1 = 0.10000000000000001
0.2 = 0.20000000000000001
0.3 = 0.29999999999999999


We're accustomed to thinking of numbers in decimal (base-10) notation, so that each fraction must be expressed as a sum of powers of 10:
$$
1 /8 = 1\cdot 10^{-1} + 2\cdot 10^{-2} + 5\cdot 10^{-3}
$$
In the familiar base-10 representation, we represent this in the familiar decimal expression: $0.125$.

Computers usually store values in binary notation, so that each number is expressed as a sum of powers of 2:
$$
1/8 = 0\cdot 2^{-1} + 0\cdot 2^{-2} + 1\cdot 2^{-3}
$$
In a base-2 representation, we can write this $0.001_2$, where the subscript 2 indicates binary notation.
The value $0.125 = 0.001_2$ happens to be one number which both binary and decimal notation can represent in a finite number of digits.

In the familiar base-10 representation of numbers, you are probably familiar with numbers that can't be expressed in a finite number of digits.
For example, dividing $1$ by $3$ gives, in standard decimal notation:
$$
1 / 3 = 0.333333333\cdots
$$
The 3s go on forever: that is, to truly represent this quotient, the number of required digits is infinite!

Similarly, there are numbers for which binary representations require an infinite number of digits.
For example:
$$
1 / 10 = 0.00011001100110011\cdots_2
$$
Just as decimal notation requires an infinite number of digits to perfectly represent $1/3$, binary notation requires an infinite number of digits to represent $1/10$.
Python internally truncates these representations at 52 bits beyond the first nonzero bit on most systems.

This rounding error for floating-point values is a necessary evil of working with floating-point numbers.
The best way to deal with it is to always keep in mind that floating-point arithmetic is approximate, and *never* rely on exact equality tests with floating-point values.

#### Iterable vs Iterator

Iterable is a collection while iterator is a stream. In iterable, all items are availble at any and all the time. And of course, you can iterate over them. However, iterator doesn't have all the items available. It gives you one item at a time on the fly and keep track of its present state. An object is called iterable if you can get an iterator for it. To convert an iterable into a iterator, pass it to `iter()` function which returns an iterator object.

In technical terms, iterable has `__iter__` method defined and iterator has `__next__` method defined for it. 

Successive calls to `next(iterator)` function return successive elements. After iterator is exhausted, it raises `StopIteration` exception. 





In [63]:
name = 'Mayank'

'__iter__' in dir(name)

True

In [64]:
a = iter(name)

'__next__' in dir(a)

True

In [65]:
next(a)

'M'

In [66]:
next(a)

'a'

In [67]:
next(a), next(a), next(a), next(a), next(a)

StopIteration: 

### PART 5

### Error handling and Exceptions

People commit all sort of mistakes when they write programs (or do anything else). If you haven't, either you are a God or you haven't tried anything. If you have come this far in your journey of learning Python, you must have seen fair amount of error messages such as `TypeError`, `SyntaxError`, `ValueError` or `NameError`. Although Python tries to help you by showing what kind of mistake you have committed (*by raising an exception*) but let's admit that seeing such message(s) isn't exactly encouraging or pleasant. 

In this section, we'll learn how to handle these errors and let our program complete gracefully regardless of whether an excception occured or not. First, let's see some examples of errors -  

In [56]:
For i in range(5)  #SyntaxError
    print(i)


SyntaxError: invalid syntax (<ipython-input-56-22fe0292925e>, line 1)

In [57]:
print(aa) #variable aa isn't defined

NameError: name 'aa' is not defined

In [58]:
1/0  #division by 0 is not allowed

ZeroDivisionError: division by zero

In [59]:
a = 0
a.value

AttributeError: 'int' object has no attribute 'value'

In [60]:
def f(a):
    a.append(2)
    
f(2)    

AttributeError: 'int' object has no attribute 'append'

Rather than showing (or seeing) these irritating messages, it'd be much nicer if we try to handle them within our program. It is not that you'll be able to avoid these errors, but at least you (or other users) will see much better output. 

Let's see how some of above programs can be rewritten to properly handle the errors ()- 

In [62]:
try:
    1/0

except ZeroDivisionError:
    print('Division by 0 is not allowed.')

Division by 0 is not allowed.


In [63]:
try:
    a = 0
    print(a.value)
    
except AttributeError as e:
    print(e)
    print("Please check for the object's attribute list")
    

'int' object has no attribute 'value'
Please check for the object's attribute list


You can see that our programs still produced errors but they were handled rather nicely. 



**Note about `SyntaxError`:**

Unlike other cases we have seen above, `SyntaxError` can't be handled. `SyntaxError` is caught when your program is parsed while errors like `AttributeError` happen when your program is exceuted. The later type of errors are called **exceptions** and we'll learn to handle these exceptions. 

Oh, just like everything else in Python, exceptions, too, are objects. 

### `try` statement


The `try` clause is followed by -

 - one or more `except` clauses (optionally followed by exactly one `else` clause)
 - (optionally) exactly one `finally` clause
 
 
The syntax is as follows - 

```Python
try:
    statement(s)
except [expression [as target]]:
    statement(s)
[else:
    statement(s)]
[finally:
    statement(s)]
```
<br>

**Note** Square brackets `[]` in documentations or syntax descriptions tell us that part enclosed by `[]` is optional. For example, if you read the [Python documentation](https://docs.python.org/3.9/library/functions.html#input) of `input()` function, you'll see that the function signature of this function is ` input([prompt])`. What it says that the parameter `prompt` is optional. So `[]` around `else` and `finally` clauses in description of `try-except-else-finally` syntax indicates that these clauses are optional. 

<br>

The `except` clauses are where exceptions are handled and therefore they are called *exception handlers*. Notice that in syntax description `except` statement is followed by `[expression [as target]]`. The `except` clause handles the exception propagated from `try`  only if exception name mathes the `expression` part in `except` statement. `expression` part may be a single class or a tuple of classes. The optional `target` is an identifier that names a variable that Python binds to the
exception object just before the exception handler executes. A handler can also obtain the current exception object by calling the `exc_info` function of module `sys`.

I can guess the above description is a bit terse. So let us go through some examples.

In [67]:
inp = input('Enter any number\n')
try:
    a =  int(inp)
    print(a)

except:
    print('Enter valid input.')
    

Enter any number
gg
Enter valid input.


Consider above example. The main part of our code resides in `try` clause. Now come to `except` clause. Any exception raised in `try` clause will be handled in `except` clause. If `try` clause doesn't raise any exception, `except` clause won't be executed. While running this code, we provided a string `gg` as input which obviously is an invalid input because it can't be converted to an integer. Because of this, an exception is raised which is caught and handled by `exception` clause. The output doesn't tell but `try` clause raised `ValueError` which we handled.

Since `exception` clause in above code doesn't have any `expression` part, it will handle every exception raised by `try` clause. Is that a good thing? Actually no.

See following example- 



In [66]:
inp = input('Enter any number\n')
try:
    a =  int(inp)
    print(aaa)

except:
    print('Enter valid number.')

Enter any number
'44'
Enter valid number.


Given our program, `'44'` was valid input and yet our program asked us to provide valid number. What went wrong?

The problem is that in 4th line of program, I mistyped `a` as `aaa` in `print` function which eventually raised the `NameError`. Since our `except` clause was designed to handle all exceptions in only one way, it handled this exception exactly the same way it earlier handled `ValueError`. Not a good thing. So what can we do to improve this? We explicitly mention the exception we want to handle in `except` clause as shown in below example -  

In [68]:
inp = input('Enter any number\n')
try:
    a =  int(inp)
    print(aaa)

except ValueError:
    print('Enter valid number.')
    
except NameError:
    print('undefined variable')
    
except:
    print("some other error")

Enter any number
54
undefined variable


This is better. 

We provided 3 `except` clauses and first 2 clause's `expression` part explicitly mentions the exception we want to handle. Since raised `NameError` matched the `expression` value of second `except` clause, this clause was executed. **Note that `except` clauses are evaluated in order**. If bare `except` clause (third clause in above example) had come before other clauses, it would have handled the `NameError`. This is not what we would have wanted. 

If you wish, you can print the error message generated by exception. This is done by providing `target` variable (in `except [expression [as target]]:` part). 

In [69]:
inp = input('Enter any number\n')

try:
    a =  int(inp)
    print(aaa)

except ValueError:
    print('Enter valid number.')
    
except NameError as err:
    print(err)
    print('undefined variable')
    
except:
    print("some other error")

Enter any number
44
name 'aaa' is not defined
undefined variable


`except` clause can handle multiple exceptions. To do this, you simply pass a tuple of exception names as an argument to `except`:

```Python
try:
    <code block>
except (NameError, ValueError):
    <code block>
```    

#### `else` clause

We mentioned we can also provide optional `else` clause. This is executed only when `try` clause is executed normally. If `try` clause raises any exception, `else` clause won't run. Also, if any of `break`, `continue` or `return` statement (if present) in `try` clause is executed, `else` clause won't run. See following examples -   

In [74]:
inp = input('Enter any number\n')
try:
    a =  int(inp)
    print(a)

except ValueError:
    print('Enter valid number.')
    
except NameError as err:
    print(err)
    print('undefined variable')
    
except:
    print("some other error")
    
else:
    print("'try' clause executed normally")
          

Enter any number
55
55
'try' clause executed normally


In [75]:
inp = input('Enter any number\n')

try:
    a =  int(inp)
    print(a)

except ValueError:
    print('Enter valid number.')
    
except NameError as err:
    print(err)
    print('undefined variable')
    
except:
    print("some other error")
    
else:
    print("'try' clause executed normally")
          

Enter any number
hfg
Enter valid number.


In [3]:
while True:

    try:    
        secret = int(input("Enter secret number"))
        if secret == 44:
            print("You are granted entry")
            break

    except ValueError:
        print("Enter valid number")
    
    else:
        print("Nothing will be printed")

Enter secret number44
You are granted entry


In [4]:
#break is executed and yet else clause ran???

try:

    while True:    
        secret = int(input("Enter secret number"))
        if secret == 44:
            print("You are granted entry")
            break

except ValueError:
    print("Enter valid number")
    
else:
    print("Nothing will be printed")

Enter secret number44
You are granted entry
Nothing will be printed


In [10]:
def f(a,b):
    try:
        a.append(b)
        print(a)    #if return a is used here, else clause won't run
    
    except AttributeError:
        print('first argument has to be a list')
    
    else:
        print('test message')

 #exception  raised so else clause won't run       

In [11]:
f(2,3)

first argument has to be a list


In [12]:
f([],3) #no exception raised, else clause executed

[3]
test message


In [15]:
def f(a,b):
    try:
        a.append(b)
        return a      #if this return executed, function will be exited 
    
    except AttributeError:
        print('first argument has to be a list')
    
    else:
        print('test message')

In [16]:
f([],2)

[2]

A note about above example, functions exit as soon as they execute `return` statement. It doesn't matter if your rest of the code was executed or not. 

In [19]:
def f(a):
    return a
    print(a*2)

In [20]:
f(32)

32

In [60]:
import sys

try:
    f = open('myfile1.txt')
    s = f.readline()
    i = int(s.strip())
except OSError as err:
    print("OS error: {0}".format(err))
except ValueError:
    print("Could not convert data to an integer.")
except:
    print("Unexpected error:", sys.exc_info()[0])
    raise

OS error: [Errno 2] No such file or directory: 'myfile1.txt'


-  Rrun this program w/o any existing 'myfile1'. It will raise `OSError`.
-  Save a text file named 'myfile' with some text content and then run the program. It will raise `ValueError`.
-  Type 'inta' in place of 'int' in `int()` in `try` clause, then run the program. It will raise `Unexpected error` in third `except` clause. Also notice the use of `sys.exc_info()` function,
-  If we omit `raise`, exception will still be raised but without any additional details. Try running the program with and w/o `raise` in third `except` clause.


#### `finally` clause

 - There can only be at most one `finally`clause. 
 - These clauses are always executed regardless of what happens in other clauses. If an exception is raised in `try` clause and is not handled, it is saved until `finally` is executed and then it is reraised. However, if `return`, `break` or `coninue` is executed in `finally` clause, the saved exception is discarded.
 - If a function has `try-finally` suit of clauses with `return` statements, the return value and any exception raised in `try` clause will be discarded and `return` value of `finally` clause will be used.  
 - If the `try` statement reaches a `break`, `continue` or `return` statement, the `finally` clause will execute just prior to the `break`, `continue` or `return` statement’s execution.


In real world applications, the `finally` clause is useful for releasing external resources (such as files or network connections), regardless of whether the use of the resource was successful.

In [21]:
try:
    1/0
    
except ZeroDivisionError:                #exception is handled
    print("division by 0 not allowed")
    
finally:
    print("this will always be executed at the end")
    

division by 0 not allowed
this will always be executed at the end


In example below,the `ZeroDivisionError` exception is raised but is not handled. So this exception is saved until the execution of `finally` clause. It is reraised after `finally` clause is executed.

In [22]:
try:
    1/0
    
except NameError:                
    print("name is not defined")
    
finally:
    print("this will always be executed at the end")
    

this will always be executed at the end


ZeroDivisionError: division by zero

In [23]:
def f(a,b):
    
    try:
        return  a/b
    finally:
        return 42
    
    

In [25]:
f(3,2)

42

In [26]:
f(1,0)

42

In [33]:
def demo1():
    try:
        raise RuntimeError #"To Force Issue"
    except:
        return 1
    else:
        return 2
    finally:
        return 3

def demo2():
    try:
        try:
            raise RuntimeError #"To Force Issue"
        except:
            return 1
        else:
            return 2
        finally:
            return 3
    except:
        print(4)
    else:
        print(5)
    finally:
        print(6)


In [34]:
demo1()

3

In [35]:
demo2()

6


3

#### Raising forced exception and `raise` statement

You can also force a specified exception to occur by using `raise` statement - 


In [37]:
raise NameError #equivalent to calling `raise NameError()`


NameError: 

In [38]:
raise NameError("Hi")

NameError: Hi

Of course, we can handle these raised errors - 

In [40]:
try:
    raise NameError("Forced error")
except NameError as e:
    print(e)
    


Forced error


We saw earlier that a bare `except` clause will handle every exception raised in `try` clause. This way, although exceptions are handled, we get no information as to what exception is handled by a bare `except` clause. We can use bare `raise` statement to re-raise the exception. Also, we can use bare `raise` to reraise any exception even though it was handled.

In [41]:
try:
    1/0
    
except ZeroDivisionError:
    print("Something bad happened")
    raise

Something bad happened


ZeroDivisionError: division by zero

In [42]:
try:
    1/0
except:
    raise

ZeroDivisionError: division by zero

#### `sys.exc_info()`

The `exc_info()` function from `sys` returns a tuple `(type, value, traceback)` regarding the most recent exception caught by an `except` clause

In [55]:
import sys

try:
    1/0
    
except:
    print(sys.exc_info()[0])
    print(sys.exc_info()[1])
    print(sys.exc_info()[2])

<class 'ZeroDivisionError'>
division by zero


In [43]:
def func():
...     raise IOError
...
>>> try:
...     func()
... except IOError as exc:
...     raise RuntimeError('Failed to open database') from exc

RuntimeError: Failed to open database

#### Defining custom exceptions

In addition to built-in exceptions, it is possible to define custom exceptions through class inheritance. For instance, if you want a special kind of `ValueError`, you can do this:

In [56]:
class CustomError(ValueError):
    pass

raise CustomError('You are wrong because I say so.')



CustomError: You are wrong because I say so.

In [57]:
a,b = 1,0

from contextlib import suppress
with suppress(ZeroDivisionError):
    print(a/b)

# error suppressed


#### Exception hiearchy

Every exception which is raised is an instance of a class derived form `BaseException`. For example, `AttributeError` is subclassed from `Exception` class which in turn is subclassed from `BaseException`. 

[Exception hiearchy](https://docs.python.org/3.9/library/exceptions.html#exception-hierarchy)